
**Audible Data** : https://1drv.ms/u/s!AiqdXCxPTydhoog8ckLN-6Cw55fzIg?e=EWgZ5d


In [1]:
import pandas as pd
import numpy as np
import re


In [ ]:
df = pd.read_excel('audible_uncleaned.xlsx')
df.shape

In [204]:
df.shape

(87489, 8)

In [205]:
df.head()

name                     author  \
0                  Geronimo Stilton #11 & #12  Writtenby:GeronimoStilton   
1                            The Burning Maze      Writtenby:RickRiordan   
2                                The Deep End       Writtenby:JeffKinney   
3                        Daughter of the Deep      Writtenby:RickRiordan   
4  The Lightning Thief: Percy Jackson, Book 1      Writtenby:RickRiordan   

                    narrator                time          releasedate  \
0      Narratedby:BillLobely   2 hrs and 20 mins  2008-04-08 00:00:00   
1   Narratedby:RobbieDaymond   13 hrs and 8 mins  2018-01-05 00:00:00   
2      Narratedby:DanRussell    2 hrs and 3 mins  2020-06-11 00:00:00   
3  Narratedby:SoneelaNankani  11 hrs and 16 mins  2021-05-10 00:00:00   
4  Narratedby:JesseBernstein              10 hrs  2010-01-13 00:00:00   

  language                          stars price  
0  English     5 out of 5 stars34 ratings   468  
1  English   4.5 out of 5 stars41 ratings   820  
2  English   4.5 out of 5 stars38 ratings   410  
3  English   4.5 out of 5 stars12 ratings   615  
4  English  4.5 out of 5 stars181 ratings   820

### summary of data

This is a dataset of the audio book platform named audible, This dataset contains data about 87489 audio books, including their ratings and prices. 

### Column descriptions

#### **Table** -> `df`: 

- `name`: the name for each of the audio books on the audible platform
- `author`: The name of each author who wrote those books, prefixed by 'writtenby' 
- `narrator`: The name of each narrator who narrate the books for audible platform, prefixed by 'Narratedby'
- `time`: The duration for each of the audio book (hour and mins)
- `release date`: The date on which the audio book was published on audible platform, date format year-month-day time
- `Language`: The spoken language for each audiobook
- `stars`: The total number of ratings and stars for each audiobook, out of 5
- `Price`: The amount for each audiobook at which the audiobooks are listed on audible


### Issues with the dataset

## 1. Dirty Data


   Table - `df`

 - `name` :

       - Some books has the version in different formats. (`Consistency`) 
            - [x] with hast-tag like "Geronimo Stilton #11 & #12". 
            - [x] "Magic Tree House Collection: Books 9-16".
            - [x] "The 39 Clues, Book 6"
            - [x] "Dungeon Item Shop: Volume 1"
            - [x] "Never Never: Part One"

        - Some books have strange letters in them: The Tyrant’s Tomb (`Accuracy`)
            - (575, 648, 736, 1056, 1148, 1431, 25849, 25858, 49372, 61642..... )
        - Some books have completely strange values like : Песочный человек (`Validity`)
            - (1090, 1341, 1342, 1343, 1348, 1349, 1636, 79304..... 79616)
        - 43 duplicated audiobooks (`Accuracy`)

 - `author` & `narrator`:
 
       - all author names has an unnecessary prefix 'WrittenBy' (`Accuracy`)
        - contains multiple writters name separated by commas or semicolon (should only be separated by commas) (`Consistency`)
        - names conatins strange character Writtenby:eOne,NicolásOluchaSánchez (`Accuracy`)
        - some writter name are completely strange Writtenby:永野裕之 (`Validity`)
        - writters name is smart reading from 11964 to 12330 (production company name) 
        - writters name is N.N from 12361 to 12422 (production company name) 
        - 12535 has writter name japanese.eyes (production company name)
        - from 12586 to 12593 has writter name Writtenby:InnovativeLanguageLearningLLC (production company name)
        - from 19384 to 19422 has writter name Writtenby:OnlineStudioProductions (production company name)
        - First name and the last are not separated with a white space, e.g. "Writtenby:NicolasGorny". instead they are separated by Capital letters (`Accuracy`)
        - There are no proper names in narrator, e.g., "Narratedby:uncredited". (could've been Nan) (`Validity`)
        - Narratedby:anonymous instead of Nan (`Validity`)

 - `time` :
    
       - Values are in incorrect format 10 hrs and 12 mins, insteasd of 10:12 (`Accuracy`)
         - wrong dtype (`Validity`)

- `releasedate`:

       - Some date format are in 2017-09-03 00:00:00 and some in 31-03-22, (`Consistency`)
         - wrong dtype (object) (`Validity`)

- `language` :
    - Some values are in title formed and some are in lower case, e.g. "English" and "german". (`Validity`)

- `stars` :

    - Not rated yet instead of Nan (`Validity`)
    - Column name can be Ratings (out of 5) so no need to make the values string by writting 4 out of 5 we can make it int (`Validity`)

- `price`
    - There is a value, "Free". This also change the dtype of the column we can change it to 0. (`Validity`)
    - Some prices are float becasue of 0s after decimal, taking more space can be converted to int (`Validity`)


## 2. Messy Data

   Table - `df`

 - `name` :

    - Some book have extra information, author name "Michael Palin: New Europe"
    - Version of the book can be separate out from book's name 

- `author` & `narrator`:
    - Some additional informations are also inclused in some values. E.g., "Writtenby:AndrewPeterson-editor,JonathanRogers,N.D.Wilson,"
    
- `stars` :
    - has two values number of ratings and ratings itself


In [206]:
df.head()

name                     author  \
0                  Geronimo Stilton #11 & #12  Writtenby:GeronimoStilton   
1                            The Burning Maze      Writtenby:RickRiordan   
2                                The Deep End       Writtenby:JeffKinney   
3                        Daughter of the Deep      Writtenby:RickRiordan   
4  The Lightning Thief: Percy Jackson, Book 1      Writtenby:RickRiordan   

                    narrator                time          releasedate  \
0      Narratedby:BillLobely   2 hrs and 20 mins  2008-04-08 00:00:00   
1   Narratedby:RobbieDaymond   13 hrs and 8 mins  2018-01-05 00:00:00   
2      Narratedby:DanRussell    2 hrs and 3 mins  2020-06-11 00:00:00   
3  Narratedby:SoneelaNankani  11 hrs and 16 mins  2021-05-10 00:00:00   
4  Narratedby:JesseBernstein              10 hrs  2010-01-13 00:00:00   

  language                          stars price  
0  English     5 out of 5 stars34 ratings   468  
1  English   4.5 out of 5 stars41 ratings   820  
2  English   4.5 out of 5 stars38 ratings   410  
3  English   4.5 out of 5 stars12 ratings   615  
4  English  4.5 out of 5 stars181 ratings   820

In [207]:
df.tail()

name                    author  \
87484  Last Days of the Bus Club    Writtenby:ChrisStewart   
87485                   The Alps   Writtenby:StephenO'Shea   
87486       The Innocents Abroad       Writtenby:MarkTwain   
87487      A Sentimental Journey  Writtenby:LaurenceSterne   
87488                     Havana   Writtenby:MarkKurlansky   

                      narrator               time          releasedate  \
87484  Narratedby:ChrisStewart  7 hrs and 34 mins  2017-09-03 00:00:00   
87485    Narratedby:RobertFass  10 hrs and 7 mins             21-02-17   
87486     Narratedby:FloGibson  19 hrs and 4 mins             30-12-16   
87487   Narratedby:AntonLesser   4 hrs and 8 mins             23-02-11   
87488   Narratedby:FleetCooper    6 hrs and 1 min  2017-07-03 00:00:00   

      language          stars price  
87484  English  Not rated yet   596  
87485  English  Not rated yet   820  
87486  English  Not rated yet   938  
87487  English  Not rated yet   680  
87488  English  Not rated yet   569

In [208]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87489 entries, 0 to 87488
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   name         87489 non-null  object
 1   author       87489 non-null  object
 2   narrator     87489 non-null  object
 3   time         87489 non-null  object
 4   releasedate  87489 non-null  object
 5   language     87489 non-null  object
 6   stars        87489 non-null  object
 7   price        87489 non-null  object
dtypes: object(8)
memory usage: 5.3+ MB


In [209]:
df.sample(5)

name  \
25230  Game of My Life: Florida Gators   
56416                Bound in Darkness   
31227    One Year to an Organized Life   
40231         A History of Air Warfare   
46032               Girls from da Hood   

                                        author                  narrator  \
25230                      Writtenby:PatDooley  Narratedby:DavidCrommett   
56416                 Writtenby:JacquelynFrank     Narratedby:RogerWayne   
31227                    Writtenby:ReginaLeeds    Narratedby:ReginaLeeds   
40231               Writtenby:JohnAndreasOlsen  Narratedby:SteveVanDoren   
46032  Writtenby:NikkiTurner,RoyGlenn,Chunichi         Narratedby:iiKane   

                     time          releasedate language          stars price  
25230    7 hrs and 2 mins             13-02-13  English  Not rated yet   668  
56416   9 hrs and 58 mins  2016-05-01 00:00:00  English  Not rated yet   938  
31227  12 hrs and 45 mins             23-04-12  English  Not rated yet   984  
40231  20 hrs and 25 mins             30-11-10  English  Not rated yet  1338  
46032   8 hrs and 32 mins             27-06-17  English  Not rated yet   820

In [210]:
df.isnull().sum()

name           0
author         0
narrator       0
time           0
releasedate    0
language       0
stars          0
price          0
dtype: int64

In [211]:
df['name'].duplicated().sum()

4722

In [212]:
df[df['name'].duplicated()]

name  \
18                      Merlin Mission Collection   
46     The Lightning Thief: Percy Jackson, Book 1   
54                      Merlin Mission Collection   
155                        Barbie - Superprincesa   
273                           Barbie - Dreamtopia   
...                                           ...   
87443       Travels with a Donkey in the Cevennes   
87456                                  Gettysburg   
87457                                        Solo   
87475                                  Wanderlust   
87486                        The Innocents Abroad   

                                           author  \
18                      Writtenby:MaryPopeOsborne   
46                          Writtenby:RickRiordan   
54                      Writtenby:MaryPopeOsborne   
155                              Writtenby:Mattel   
273    Writtenby:Mattel,MartaCisaMuñoz-traductor   
...                                           ...   
87443              Writtenby:RobertLouisStevenson   
87456                        Writtenby:JeffShaara   
87457                          Writtenby:PenHadow   
87475                    Writtenby:ElisabethEaves   
87486                         Writtenby:MarkTwain   

                             narrator                time  \
18         Narratedby:MaryPopeOsborne  10 hrs and 18 mins   
46             Narratedby:WalterLewis   3 hrs and 51 mins   
54         Narratedby:MaryPopeOsborne  13 hrs and 24 mins   
155    Narratedby:VanessaPérezJurado             29 mins   
273          Narratedby:MiriamMonlleo             21 mins   
...                               ...                 ...   
87443          Narratedby:DenisLawson   2 hrs and 51 mins   
87456        Narratedby:RobertsonDean    1 hr and 12 mins   
87457             Narratedby:PenHadow    3 hrs and 5 mins   
87475          Narratedby:ErinBennett   9 hrs and 51 mins   
87486            Narratedby:FloGibson   19 hrs and 4 mins   

               releasedate language                       stars price  
18     2017-02-05 00:00:00  English  5 out of 5 stars11 ratings  1256  
46                13-01-10  English   4 out of 5 stars4 ratings   615  
54     2017-02-05 00:00:00  English   5 out of 5 stars8 ratings  1256  
155    2022-07-03 00:00:00  spanish               Not rated yet    38  
273               31-01-22  catalan               Not rated yet    38  
...                    ...      ...                         ...   ...  
87443             13-05-08  English               Not rated yet   569  
87456             26-03-07  English               Not rated yet   200  
87457             14-02-05  English               Not rated yet   615  
87475             26-02-14  English               Not rated yet   668  
87486             30-12-16  English               Not rated yet   938  

[4722 rows x 8 columns]

In [213]:
df[df['name'] == 'Merlin Mission Collection']

name                     author  \
8   Merlin Mission Collection  Writtenby:MaryPopeOsborne   
18  Merlin Mission Collection  Writtenby:MaryPopeOsborne   
54  Merlin Mission Collection  Writtenby:MaryPopeOsborne   

                      narrator                time          releasedate  \
8   Narratedby:MaryPopeOsborne  10 hrs and 56 mins  2017-02-05 00:00:00   
18  Narratedby:MaryPopeOsborne  10 hrs and 18 mins  2017-02-05 00:00:00   
54  Narratedby:MaryPopeOsborne  13 hrs and 24 mins  2017-02-05 00:00:00   

   language                       stars price  
8   English   5 out of 5 stars5 ratings  1256  
18  English  5 out of 5 stars11 ratings  1256  
54  English   5 out of 5 stars8 ratings  1256

In [214]:
	
df[df['name'] == 'The Innocents Abroad']

name               author                  narrator  \
86364  The Innocents Abroad  Writtenby:MarkTwain     Narratedby:RobinField   
87017  The Innocents Abroad  Writtenby:MarkTwain  Narratedby:GroverGardner   
87486  The Innocents Abroad  Writtenby:MarkTwain      Narratedby:FloGibson   

                     time          releasedate language  \
86364   24 hrs and 4 mins             20-09-10  English   
87017  18 hrs and 13 mins  2011-09-09 00:00:00  English   
87486   19 hrs and 4 mins             30-12-16  English   

                             stars price  
86364     5 out of 5 stars1 rating  1640  
87017  4.5 out of 5 stars2 ratings   937  
87486                Not rated yet   938

Narrator can be multiple for the same audio book 

In [215]:
df.duplicated(subset=['name', 'narrator','time','releasedate']).sum()

43

In [216]:
df[df.duplicated(subset=['name', 'narrator','time','releasedate'])]

name  \
26848                  The Woman They Could Not Silence   
33922       What I Talk About When I Talk About Running   
35965                                      Finding Flow   
36606                      Konzentriert lernen mit ADHS   
38021                        When Women Ruled the World   
41111                   Aurangzeb: The Man and the Myth   
41680                              Kitchen Confidential   
54339                                             Freed   
57871                     Bulls, Bears and Other Beasts   
58570                           The Psychology of Money   
59166                               Poirot Investigates   
60840                        The Picture of Dorian Gray   
61863                      The Courage to Be Disliked    
61881                                          Educated   
61884                                  Liminal Thinking   
62265                    The Almanack of Naval Ravikant   
62269                                        Hyperfocus   
62272                                     Atomic Habits   
62279                             How to Talk to Anyone   
62281                        The Richest Man in Babylon   
62284                                              Flow   
62305                             Becoming Supernatural   
62348                            Emotional Intelligence   
62362                            How to Train Your Mind   
62404                          The Happiness Hypothesis   
62409         How to Talk to Anybody, Anytime, Anywhere   
62415                               The Compound Effect   
62460       Cracking the Code to a Successful Interview   
64428                                     Wings of Fire   
64440                   How to Avoid a Climate Disaster   
64448                                       Che Guevara   
64451            The Story of My Experiments with Truth   
64722                   100 quotes by Benjamin Franklin   
64799                                          Figuring   
67521                                         Waking Up   
67540                                     Spirit Babies   
79107                              The Molecule of More   
79753  Harry Potter and the Philosopher's Stone, Book 1   
79754  Harry Potter and the Prisoner of Azkaban, Book 3   
79756                  Mahayoddha Kalki: Sword of Shiva   
79757                                            Mythos   
79801                     He Who Fights with Monsters 5   
80078                               Path of Destruction   

                                                  author  \
26848                                Writtenby:KateMoore   
33922                           Writtenby:HarukiMurakami   
35965                   Writtenby:MihalyCsikszentmihalyi   
36606                             Writtenby:SaraDörwald   
38021                               Writtenby:KaraCooney   
41111                           Writtenby:AudreyTrushcke   
41680                          Writtenby:AnthonyBourdain   
54339                                Writtenby:E.L.James   
57871                              Writtenby:SantoshNair   
58570                             Writtenby:MorganHousel   
59166                           Writtenby:AgathaChristie   
60840                               Writtenby:OscarWilde   
61863               Writtenby:FumitakeKoga,IchiroKishimi   
61881                             Writtenby:TaraWestover   
61884      Writtenby:DaveGray,RichardSaulWurman-foreword   
62265                 Writtenby:EricJorgenson,TimFerriss   
62269                              Writtenby:ChrisBailey   
62272                               Writtenby:JamesClear   
62279                              Writtenby:LeilLowndes   
62281                           Writtenby:GeorgeS.Clason   
62284              Writtenby:MihalyCsikszentmihalyiPh.D.   
62305                              Writtenby:JoeDispenza   
62348                            Writtenby:DanielGoleman   
62362         

In [217]:
df[df['name'] == 'Freed']

name               author                  narrator  \
54311  Freed  Writtenby:E.L.James  Narratedby:ZacharyWebber   
54339  Freed  Writtenby:E.L.James  Narratedby:ZacharyWebber   

                     time          releasedate language  \
54311  27 hrs and 15 mins  2021-01-06 00:00:00  English   
54339  27 hrs and 15 mins  2021-01-06 00:00:00  English   

                              stars price  
54311  4.5 out of 5 stars12 ratings  1093  
54339                 Not rated yet  1382

In [218]:
df.describe() 

name                            author  \
count            87489                             87489   
unique           82767                             48374   
top     The Art of War  Writtenby:矢島雅弘,石橋遊   
freq                20                               874   

                    narrator    time releasedate language          stars  \
count                  87489   87489       87489    87489          87489   
unique                 29717    2284        5059       36            665   
top     Narratedby:anonymous  2 mins    16-05-18  English  Not rated yet   
freq                    1034     372         773    61884          72417   

        price  
count   87489  
unique   1011  
top       586  
freq     5533

In [219]:
df[df['name'] == 'The Art of War']

name                                         author  \
15112  The Art of War                               Writtenby:SunTzu   
27377  The Art of War                               Writtenby:SunTzu   
27977  The Art of War                               Writtenby:SunTzu   
27995  The Art of War                               Writtenby:SunTzu   
29508  The Art of War                               Writtenby:SunTzu   
29673  The Art of War                               Writtenby:SunTzu   
29777  The Art of War       Writtenby:ThomasCleary-translator,SunTzu   
30158  The Art of War                               Writtenby:SunTzu   
38116  The Art of War  Writtenby:IntroductionbyStefanRudnicki,SunTzu   
38227  The Art of War                               Writtenby:SunTzu   
38355  The Art of War                               Writtenby:SunTzu   
38779  The Art of War                               Writtenby:SunTzu   
38849  The Art of War                               Writtenby:SunTzu   
38961  The Art of War                               Writtenby:SunTzu   
39224  The Art of War         Writtenby:LionelGilestranslator,SunTzu   
40227  The Art of War                               Writtenby:SunTzu   
40299  The Art of War                               Writtenby:SunTzu   
41246  The Art of War                               Writtenby:SunTzu   
41545  The Art of War                               Writtenby:SunTzu   
65687  The Art of War                               Writtenby:SunTzu   

                                 narrator               time  \
15112           Narratedby:PhillipJMather    1 hr and 8 mins   
27377              Narratedby:AidanGillen    1 hr and 7 mins   
27977                Narratedby:MelFoster   1 hr and 23 mins   
27995    Narratedby:DenmaTranslationGroup  3 hrs and 51 mins   
29508                 Narratedby:AlecSand   1 hr and 46 mins   
29673  Narratedby:DonHagen,VictoriaGordon  6 hrs and 29 mins   
29777               Narratedby:LloydJames   1 hr and 54 mins   
30158               Narratedby:MikeBorris   1 hr and 48 mins   
38116       Narratedby:RonSilver,B.D.Wong  2 hrs and 15 mins   
38227             Narratedby:SethThompson  6 hrs and 35 mins   
38355             Narratedby:GrainneRegan    1 hr and 9 mins   
38779            Narratedby:KennethElliot   1 hr and 22 mins   
38849              Narratedby:DanStrutzel   1 hr and 38 mins   
38961             Narratedby:GabraZackman   1 hr and 42 mins   
39224                Narratedby:DenisDaly   1 hr and 23 mins   
40227             Narratedby:JimDonaldson   1 hr and 13 mins   
40299      Narratedby:ElizabethAnneHarris   1 hr and 12 mins   
41246               Narratedby:SimonVance   2 hrs and 6 mins   
41545             Narratedby:MikeVendetti   1 hr and 12 mins   
65687                Narratedby:PaulSpera    1 hr and 8 mins   

               releasedate language                        stars price  
15112  2015-11-12 00:00:00  English                Not rated yet   333  
27377             31-03-15  English  4 out of 5 stars511 ratings   333  
27977  2016-06-12 00:00:00  English    4 out of 5 stars4 ratings   333  
27995  2016-12-08 00:00:00  English     5 out of 5 stars1 rating   501  
29508             18-09-09  English    4 out of 5 stars4 ratings   268  
29673             22-09-11  English  3.5 out of 5 stars3 ratings   668  
29777  2018-03-04 00:00:00  English    5 out of 5 stars6 ratings   377  
30158  2011-01-11 00:00:00  English     4 out of 5 stars1 rating   333  
38116             16-12-99  English     5 out of 5 stars1 rating   333  
38227  2020-01-04 00:00:00  English    2 out of 5 stars2 ratings   258  
38355  2018-12-12 00:00:00  English                Not rated yet    65  
38779  2020-04-03 00:00:00  English  4.5 out of 5 stars2 ratings    32  
38849             28-04-21  English                Not rated yet  1003  
38961  2020-07-01 00:00:00  English                Not rated yet   702  
39224             31-07-17  English                Not rated yet

In [220]:
new_df = df.copy()
def foo(x):
    x = str(x)
    return x.split(':')
new_df['name'] = new_df['name'].apply(foo)

new_df.isnull().sum()

name           0
author         0
narrator       0
time           0
releasedate    0
language       0
stars          0
price          0
dtype: int64

In [221]:
# messy data : removing authors name from the name of the book


def foo(x):
    if len(x['name']) <= 1:
        return x
    x['name'] = [i.strip() for i in x['name']]
    writter = x['author'].split(':')[1]
    for ndx, i in enumerate(x['name']):
        if i.replace(' ','') == writter:
            x['name'].remove(i)
            
            return x
    return x


new_df = new_df.apply(foo, axis=1)



In [222]:
# testing
new_df[new_df['name'].str.get(0) == 'The seven voyages of Sindbad the Sailor']

name                author  \
1053  [The seven voyages of Sindbad the Sailor]  Writtenby:AndrewLang   

                        narrator              time releasedate language  \
1053  Narratedby:JürgenFritsche  1 hr and 49 mins    16-07-21  English   

              stars price  
1053  Not rated yet   267

In [223]:
new_df[new_df['name'].str.get(0) == 'New Europe']

name                  author                 narrator  \
85941  [New Europe]  Writtenby:MichaelPalin  Narratedby:MichaelPalin   
86512  [New Europe]  Writtenby:MichaelPalin  Narratedby:MichaelPalin   

                     time          releasedate language          stars price  
85941  11 hrs and 26 mins  2012-11-10 00:00:00  English  Not rated yet   748  
86512   6 hrs and 50 mins             17-10-07  English  Not rated yet   607

In [224]:
new_df['name'] = new_df['name'].str.join(' ')

new_df['name'] = new_df['name'].str.split(' ')

new_df

name  \
0                       [Geronimo, Stilton, #11, &, #12]   
1                                   [The, Burning, Maze]   
2                                       [The, Deep, End]   
3                              [Daughter, of, the, Deep]   
4      [The, Lightning, Thief, Percy, Jackson,, Book, 1]   
...                                                  ...   
87484                   [Last, Days, of, the, Bus, Club]   
87485                                        [The, Alps]   
87486                           [The, Innocents, Abroad]   
87487                          [A, Sentimental, Journey]   
87488                                           [Havana]   

                          author                   narrator  \
0      Writtenby:GeronimoStilton      Narratedby:BillLobely   
1          Writtenby:RickRiordan   Narratedby:RobbieDaymond   
2           Writtenby:JeffKinney      Narratedby:DanRussell   
3          Writtenby:RickRiordan  Narratedby:SoneelaNankani   
4          Writtenby:RickRiordan  Narratedby:JesseBernstein   
...                          ...                        ...   
87484     Writtenby:ChrisStewart    Narratedby:ChrisStewart   
87485    Writtenby:StephenO'Shea      Narratedby:RobertFass   
87486        Writtenby:MarkTwain       Narratedby:FloGibson   
87487   Writtenby:LaurenceSterne     Narratedby:AntonLesser   
87488    Writtenby:MarkKurlansky     Narratedby:FleetCooper   

                     time          releasedate language  \
0       2 hrs and 20 mins  2008-04-08 00:00:00  English   
1       13 hrs and 8 mins  2018-01-05 00:00:00  English   
2        2 hrs and 3 mins  2020-06-11 00:00:00  English   
3      11 hrs and 16 mins  2021-05-10 00:00:00  English   
4                  10 hrs  2010-01-13 00:00:00  English   
...                   ...                  ...      ...   
87484   7 hrs and 34 mins  2017-09-03 00:00:00  English   
87485   10 hrs and 7 mins             21-02-17  English   
87486   19 hrs and 4 mins             30-12-16  English   
87487    4 hrs and 8 mins             23-02-11  English   
87488     6 hrs and 1 min  2017-07-03 00:00:00  English   

                               stars price  
0         5 out of 5 stars34 ratings   468  
1       4.5 out of 5 stars41 ratings   820  
2       4.5 out of 5 stars38 ratings   410  
3       4.5 out of 5 stars12 ratings   615  
4      4.5 out of 5 stars181 ratings   820  
...                              ...   ...  
87484                  Not rated yet   596  
87485                  Not rated yet   820  
87486                  Not rated yet   938  
87487                  Not rated yet   680  
87488                  Not rated yet   569  

[87489 rows x 8 columns]

In [225]:
# messy data : Extracting version from the name of the book

import re
import numpy as np



New_book_name = pd.DataFrame({
    'name': []
})


def convert_to_int(book_name):
    
    for ndx, i in enumerate(book_name):
        try:
            book_name[ndx] = str(w2n.word_to_num(i))
        except:
            book_name[ndx] = str(i)
    
    return ' '.join(book_name)

def extract_version(book_name):
    global New_book_name
    book_name = convert_to_int(book_name)
    regex = r'(?:\bBook\b|\bbook\b|\bLessons\b|\bBooks\b|\bPart\b|#|\bEdition\b|\bLevel\b|\bVol\b|\bvol\b|\bVolume\b|\bvolume\b|Tome \d+|Tome \d+-\d+|Tome \d+,\d+,\d+|Tome \d+,\d+|Tome \d+,\d+,\d+,\d+|#\s*\d+(?:\s*&\s*#\d+)?|Tome \d+,\d+)\s*([#\d\s\band\b|&-]+\d+)'
    match = re.search(regex, book_name)
    versions = []

    if match:
        while match:
            if match.group(1).strip() == '':
                New_book_name = pd.concat([New_book_name, pd.DataFrame({'name': [book_name]})], ignore_index=True, axis=0)
                return np.nan
            versions.append(match.group(0))
            book_name = book_name.replace(match.group(0), '', 1)
            match = re.search(regex, book_name)

    else:
        New_book_name = pd.concat([New_book_name, pd.DataFrame({'name': [book_name]})], ignore_index=True, axis=0)
        return np.nan

    New_book_name = pd.concat([New_book_name, pd.DataFrame({'name': [book_name]})], ignore_index=True, axis=0)
    return versions

new_df['version'] = new_df['name'].apply(extract_version)

In [226]:
# testing version code
new_df[~new_df['version'].isnull()][['version']].sample(10)

version
77032                           [Book 3]
12840  [Level 1, Volume 1, Lessons 1-25]
12958                         [Volume 2]
12960                          [Level 1]
57985                            [#2020]
38268                         [Volume 3]
32566                           [Book 4]
50605                           [Part 2]
13412                     [Lessons 1-25]
77158                          [Book 11]

In [227]:
# testing version code
new_df[~new_df['version'].isnull()].to_excel('version.xlsx', index=False)

In [228]:
new_df[~new_df['version'].isnull()].head(20)

name  \
0                     [Geronimo, Stilton, #11, &, #12]   
4    [The, Lightning, Thief, Percy, Jackson,, Book, 1]   
10     [The, Titan's, Curse, Percy, Jackson,, Book, 3]   
11       [Magic, Tree, House, Collection, Books, 9-16]   
12        [Magic, Tree, House, Collection, Books, 1-8]   
15                  [Geronimo, Stilton, #13, and, #14]   
21         [Eldest, The, Inheritance, Cycle,, Book, 2]   
26      [Magic, Tree, House, Collection, Books, 25-32]   
27                  [Geronimo, Stilton, #20, and, #21]   
30                            [Geronimo, Stilton, #17]   
36                  [Geronimo, Stilton, #15, and, #16]   
46   [The, Lightning, Thief, Percy, Jackson,, Book, 1]   
66                          [The, 39, Clues,, Book, 6]   
103                         [The, 39, Clues,, Book, 4]   
104                         [The, 39, Clues,, Book, 3]   
137            [The, 39, Clues, Unstoppable,, Book, 3]   
329                         [The, 39, Clues,, Book, 5]   
334                         [The, 39, Clues,, Book, 9]   
350                        [The, 39, Clues,, Book, 10]   
353  [Captain, Underpants, and, the, Big,, Bad, Bat...   

                               author  \
0           Writtenby:GeronimoStilton   
4               Writtenby:RickRiordan   
10              Writtenby:RickRiordan   
11          Writtenby:MaryPopeOsborne   
12          Writtenby:MaryPopeOsborne   
15          Writtenby:GeronimoStilton   
21       Writtenby:ChristopherPaolini   
26          Writtenby:MaryPopeOsborne   
27          Writtenby:GeronimoStilton   
30          Writtenby:GeronimoStilton   
36          Writtenby:GeronimoStilton   
46              Writtenby:RickRiordan   
66               Writtenby:JudeWatson   
103              Writtenby:JudeWatson   
104           Writtenby:PeterLerangis   
137       Writtenby:NatalieStandiford   
329           Writtenby:PatrickCarman   
334            Writtenby:LindaSuePark   
350  Writtenby:MargaretPetersonHaddix   
353               Writtenby:DavPilkey   

                                             narrator                time  \
0                               Narratedby:BillLobely   2 hrs and 20 mins   
4                           Narratedby:JesseBernstein              10 hrs   
10                          Narratedby:JesseBernstein   8 hrs and 48 mins   
11                         Narratedby:MaryPopeOsborne   5 hrs and 23 mins   
12                         Narratedby:MaryPopeOsborne     6 hrs and 1 min   
15                              Narratedby:BillLobley   2 hrs and 25 mins   
21                            Narratedby:GerrardDoyle  23 hrs and 29 mins   
26                         Narratedby:MaryPopeOsborne    7 hrs and 9 mins   
27                              Narratedby:BillLobley   2 hrs and 32 mins   
30                              Narratedby:BillLobley   2 hrs and 28 mins   
36                              Narratedby:BillLobley   2 hrs and 43 mins   
46                             Narratedby:WalterLewis   3 hrs and 51 mins   
66                              Narratedby:DavidPittu   5 hrs and 10 mins   
103                             Narratedby:DavidPittu   4 hrs and 37 mins   
104                             Narratedby:DavidPittu   3 hrs and 51 mins   
137                             Narratedby:DavidPittu   4 hrs and 49 mins   
329                             Narratedby:DavidPittu               4 hrs   
334                             Narratedby:DavidPittu   4 hrs and 45 mins   
350                             Narratedby:DavidPittu    7 hrs and 8 mins   
353  Narratedby:LenForgione,BenD'Amico,I’keMitchell    1 hr and 23 mins   

             releasedate language                          stars price  \
0    2008-04-08 00:00:00  English     5 out of 5 stars34 ratings   468   
4    2010-01-13 00:00:00  English  4.5 out of 5 stars181 ratings   820   
10              14-01-10  English  4.5 out of 5 stars130 ratings   820   
11              24-08-11  English      5 out of 5 stars

In [229]:
New_book_name, new_df.shape

(                                      name
 0                        Geronimo Stilton 
 1                         The Burning Maze
 2                             The Deep End
 3                     Daughter of the Deep
 4      The Lightning Thief Percy Jackson, 
 ...                                    ...
 87484            Last Days of the Bus Club
 87485                             The Alps
 87486                 The Innocents Abroad
 87487                A Sentimental Journey
 87488                               Havana
 
 [87489 rows x 1 columns],
 (87489, 9))

In [230]:
New_book_name.iloc[4], new_df.iloc[4] 

(name    The Lightning Thief Percy Jackson, 
 Name: 4, dtype: object,
 name           [The, Lightning, Thief, Percy, Jackson,, Book, 1]
 author                                     Writtenby:RickRiordan
 narrator                               Narratedby:JesseBernstein
 time                                                      10 hrs
 releasedate                                  2010-01-13 00:00:00
 language                                                 English
 stars                              4.5 out of 5 stars181 ratings
 price                                                        820
 version                                                 [Book 1]
 Name: 4, dtype: object)

In [231]:
New_book_name

name
0                        Geronimo Stilton 
1                         The Burning Maze
2                             The Deep End
3                     Daughter of the Deep
4      The Lightning Thief Percy Jackson, 
...                                    ...
87484            Last Days of the Bus Club
87485                             The Alps
87486                 The Innocents Abroad
87487                A Sentimental Journey
87488                               Havana

[87489 rows x 1 columns]

In [232]:
new_df.name

0                         [Geronimo, Stilton, #11, &, #12]
1                                     [The, Burning, Maze]
2                                         [The, Deep, End]
3                                [Daughter, of, the, Deep]
4        [The, Lightning, Thief, Percy, Jackson,, Book, 1]
                               ...                        
87484                     [Last, Days, of, the, Bus, Club]
87485                                          [The, Alps]
87486                             [The, Innocents, Abroad]
87487                            [A, Sentimental, Journey]
87488                                             [Havana]
Name: name, Length: 87489, dtype: object

In [233]:
new_df['name'] = New_book_name['name']

In [234]:
new_df['name'] # version removed from name column

0                          Geronimo Stilton 
1                           The Burning Maze
2                               The Deep End
3                       Daughter of the Deep
4        The Lightning Thief Percy Jackson, 
                        ...                 
87484              Last Days of the Bus Club
87485                               The Alps
87486                   The Innocents Abroad
87487                  A Sentimental Journey
87488                                 Havana
Name: name, Length: 87489, dtype: object

In [235]:

import re
from word2number import w2n

def convert_to_int(book_name):
    for ndx, i in enumerate(book_name):
        try:
            book_name[ndx] = str(w2n.word_to_num(i))
        except:
            pass

    return book_name

def extract_versions(book_name):
    versions = []
    regex = r'(?:\bBook\b|\bbook\b|\bLessons\b|\bBooks\b|\bPart\b|#|\bEdition\b|\bLevel\b|\bVol\b|\bvol\b|\bVolume\b|\bvolume\b|Tome \d+|Tome \d+-\d+|Tome \d+,\d+,\d+|Tome \d+,\d+|Tome \d+,\d+,\d+,\d+|#\s*\d+(?:\s*&\s*#\d+)?|Tome \d+,\d+)\s*([#\d\s\band\b|&-]+\d+)'
    match = re.search(regex, book_name)
    while match:
        version = match.group(1).strip()
        versions.append(version)
        book_name = book_name.replace(match.group(0), '', 1)
        match = re.search(regex, book_name)
    return versions


book_name = "The Princess in Black, : The Princess in Black; The Princess in Black and the Perfect Princess Book-Red Book 1 Level 1 Volume 1" 
versions = extract_versions(book_name)
print(versions)

book_name = str(book_name).strip().split()

book_name = convert_to_int(book_name)

print(book_name)

book_name = ' '.join(book_name)

version = extract_version(book_name)
if version:
    print(f"Version: {version}")
else:
    print("No version found.")

['1', '1', '1']
['The', 'Princess', 'in', 'Black,', ':', 'The', 'Princess', 'in', 'Black;', 'The', 'Princess', 'in', 'Black', 'and', 'the', 'Perfect', 'Princess', 'Book-Red', 'Book', '1', 'Level', '1', 'Volume', '1']
Version: ['Book 1', 'Level 1', 'Volume 1']


In [237]:
import re

def check_lowercase_after_hyphen(string):


    regex = r'(\b\w+\b)\s*-\s*([a-z]{4,})'
    matches = re.findall(regex, string)

    if matches:
        return matches
    else:
        return None
    
new_df['Extra_Staff'] = new_df['author'].apply(check_lowercase_after_hyphen)

# removing the extra staff from the author column

def remove_extra_staff(x):
    if x['Extra_Staff']:
        for i in x['Extra_Staff']:
            
            x['author'] = x['author'].replace(' '.join(i).replace(' ','-'), '')
    return x

    
new_df = new_df.apply(remove_extra_staff, axis=1)

In [238]:
#testing
new_df[new_df['Extra_Staff'].notnull()] # maybe we need to run this function, check_lowercase_after_hyphen, again after removing unwanted characters from the author column


name  \
109                  Danger at Dead Man's Pass   
117                 Barbie - Aventura espacial   
131                               3 Lost Seeds   
142    The Christmasaurus and the Naughty List   
161     Scary Stories for Young Foxes The City   
...                                        ...   
86853                           St. Louis Noir   
86854                            Brooklyn Noir   
86893                             Oakland Noir   
86905                         Mississippi Noir   
87350                            Wild Yosemite   

                                   author  \
109     Writtenby:M.G.Leonard,SamSedgman,   
117        Writtenby:Mattel,MartaCisaMuñ   
131              Writtenby:StephieMorton,   
142                Writtenby:TomFletcher,   
161    Writtenby:ChristianMcKayHeidicker,   
...                                   ...   
86853           Writtenby:variousauthors,   
86854                          Writtenby:   
86893                         Writtenby:,   
86905           Writtenby:variousauthors,   
87350                   Writtenby:SusanM.   

                                                narrator                time  \
109                                 Narratedby:JotDavies   6 hrs and 48 mins   
117                             Narratedby:MiriamMonlleo             12 mins   
131                            Narratedby:StevenJayCohen             13 mins   
142                   Narratedby:PaulShelley,TomFletcher   6 hrs and 15 mins   
161                   Narratedby:ChristianMcKayHeidicker   9 hrs and 13 mins   
...                                                  ...                 ...   
86853  Narratedby:RebeccaGibel,AdenreleOjo,KevinKenerly,   7 hrs and 37 mins   
86854    Narratedby:MichaelBraun,AdamChase,KarenChilton,  11 hrs and 20 mins   
86893  Narratedby:TraberBurns,KeithSzarabajka,Adenrel...    8 hrs and 2 mins   
86905    Narratedby:AdenreleOjo,RebeccaGibel,LloydJames,   8 hrs and 49 mins   
87350                              Narratedby:ErikDavies    9 hrs and 4 mins   

               releasedate language          stars price version  \
109               16-09-21  English  Not rated yet   323     NaN   
117    2022-02-02 00:00:00  catalan  Not rated yet    76     NaN   
131    2022-08-02 00:00:00  English  Not rated yet   234     NaN   
142               14-10-21  English  Not rated yet   615     NaN   
161               28-09-21  English  Not rated yet   703     NaN   
...                    ...      ...            ...   ...     ...   
86853  2016-02-08 00:00:00  English  Not rated yet   820     NaN   
86854             25-11-14  English  Not rated yet   836     NaN   
86893  2017-04-04 00:00:00  English  Not rated yet   820     NaN   
86905  2016-02-08 00:00:00  English  Not rated yet   820     NaN   
87350  2013-04-02 00:00:00  English  Not rated yet   668     NaN   

                                            Extra_Staff  
109                     [(ElisaPaganelli, illustrator)]  
117                                   [(oz, traductor)]  
131                         [(NicoleWong, illustrator)]  
142                       [(ShaneDevries, illustrator)]  
161                            [(JunyiWu, illustrator)]  
...                                                 ...  
86853                         [(ScottPhillips, editor)]  
86854                         [(TimMcLoughlin, editor)]  
86893  [(JerryThompson, editor), (EddieMuller, editor)]  
86905                           [(TomFranklin, editor)]  
87350                                [(Neider, editor)]  

[2576 rows x 10 columns]

In [239]:
# Messydata: removing number of rating and rating itself from the stars column


def find_total_ratings(x):
    if x['stars'] == 'Not rated yet':
        x['total_ratings'] = 0
        x['stars'] = np.nan
        return x

    else:
        tol_rattings_pattern = re.compile(r"[\d]+\sratings")
        result = re.findall(tol_rattings_pattern, x['stars'])

        try:
            # if we find the total ratings, then convert to int and return it
            if len(result) > 0:
                x['total_ratings'] = int(result[0].split(" ")[0])
                return x

            else:
                return x
        except Exception as e:
            
            return x


new_df = new_df.apply(find_total_ratings, axis=1)

In [240]:
new_df['Average Ratings'] = new_df["stars"].apply(lambda x: float(x.split(" ")[0]) if not pd.isna(x) else np.nan)

In [241]:
new_df

Extra_Staff                     author language  \
0            None  Writtenby:GeronimoStilton  English   
1            None      Writtenby:RickRiordan  English   
2            None       Writtenby:JeffKinney  English   
3            None      Writtenby:RickRiordan  English   
4            None      Writtenby:RickRiordan  English   
...           ...                        ...      ...   
87484        None     Writtenby:ChrisStewart  English   
87485        None    Writtenby:StephenO'Shea  English   
87486        None        Writtenby:MarkTwain  English   
87487        None   Writtenby:LaurenceSterne  English   
87488        None    Writtenby:MarkKurlansky  English   

                                      name                   narrator price  \
0                        Geronimo Stilton       Narratedby:BillLobely   468   
1                         The Burning Maze   Narratedby:RobbieDaymond   820   
2                             The Deep End      Narratedby:DanRussell   410   
3                     Daughter of the Deep  Narratedby:SoneelaNankani   615   
4      The Lightning Thief Percy Jackson,   Narratedby:JesseBernstein   820   
...                                    ...                        ...   ...   
87484            Last Days of the Bus Club    Narratedby:ChrisStewart   596   
87485                             The Alps      Narratedby:RobertFass   820   
87486                 The Innocents Abroad       Narratedby:FloGibson   938   
87487                A Sentimental Journey     Narratedby:AntonLesser   680   
87488                               Havana     Narratedby:FleetCooper   569   

               releasedate                          stars                time  \
0      2008-04-08 00:00:00     5 out of 5 stars34 ratings   2 hrs and 20 mins   
1      2018-01-05 00:00:00   4.5 out of 5 stars41 ratings   13 hrs and 8 mins   
2      2020-06-11 00:00:00   4.5 out of 5 stars38 ratings    2 hrs and 3 mins   
3      2021-05-10 00:00:00   4.5 out of 5 stars12 ratings  11 hrs and 16 mins   
4      2010-01-13 00:00:00  4.5 out of 5 stars181 ratings              10 hrs   
...                    ...                            ...                 ...   
87484  2017-09-03 00:00:00                            NaN   7 hrs and 34 mins   
87485             21-02-17                            NaN   10 hrs and 7 mins   
87486             30-12-16                            NaN   19 hrs and 4 mins   
87487             23-02-11                            NaN    4 hrs and 8 mins   
87488  2017-07-03 00:00:00                            NaN     6 hrs and 1 min   

       total_ratings      version  Average Ratings  
0               34.0  [#11 & #12]              5.0  
1               41.0          NaN              4.5  
2               38.0          NaN              4.5  
3               12.0          NaN              4.5  
4              181.0     [Book 1]              4.5  
...              ...          ...              ...  
87484            0.0          NaN              NaN  
87485            0.0          NaN              NaN  
87486            0.0          NaN              NaN  
87487            0.0          NaN              NaN  
87488            0.0          NaN              NaN  

[87489 rows x 12 columns]

In [242]:
new_df.drop(columns=['stars'],inplace = True)
new_df

Extra_Staff                     author language  \
0            None  Writtenby:GeronimoStilton  English   
1            None      Writtenby:RickRiordan  English   
2            None       Writtenby:JeffKinney  English   
3            None      Writtenby:RickRiordan  English   
4            None      Writtenby:RickRiordan  English   
...           ...                        ...      ...   
87484        None     Writtenby:ChrisStewart  English   
87485        None    Writtenby:StephenO'Shea  English   
87486        None        Writtenby:MarkTwain  English   
87487        None   Writtenby:LaurenceSterne  English   
87488        None    Writtenby:MarkKurlansky  English   

                                      name                   narrator price  \
0                        Geronimo Stilton       Narratedby:BillLobely   468   
1                         The Burning Maze   Narratedby:RobbieDaymond   820   
2                             The Deep End      Narratedby:DanRussell   410   
3                     Daughter of the Deep  Narratedby:SoneelaNankani   615   
4      The Lightning Thief Percy Jackson,   Narratedby:JesseBernstein   820   
...                                    ...                        ...   ...   
87484            Last Days of the Bus Club    Narratedby:ChrisStewart   596   
87485                             The Alps      Narratedby:RobertFass   820   
87486                 The Innocents Abroad       Narratedby:FloGibson   938   
87487                A Sentimental Journey     Narratedby:AntonLesser   680   
87488                               Havana     Narratedby:FleetCooper   569   

               releasedate                time  total_ratings      version  \
0      2008-04-08 00:00:00   2 hrs and 20 mins           34.0  [#11 & #12]   
1      2018-01-05 00:00:00   13 hrs and 8 mins           41.0          NaN   
2      2020-06-11 00:00:00    2 hrs and 3 mins           38.0          NaN   
3      2021-05-10 00:00:00  11 hrs and 16 mins           12.0          NaN   
4      2010-01-13 00:00:00              10 hrs          181.0     [Book 1]   
...                    ...                 ...            ...          ...   
87484  2017-09-03 00:00:00   7 hrs and 34 mins            0.0          NaN   
87485             21-02-17   10 hrs and 7 mins            0.0          NaN   
87486             30-12-16   19 hrs and 4 mins            0.0          NaN   
87487             23-02-11    4 hrs and 8 mins            0.0          NaN   
87488  2017-07-03 00:00:00     6 hrs and 1 min            0.0          NaN   

       Average Ratings  
0                  5.0  
1                  4.5  
2                  4.5  
3                  4.5  
4                  4.5  
...                ...  
87484              NaN  
87485              NaN  
87486              NaN  
87487              NaN  
87488              NaN  

[87489 rows x 11 columns]

In [247]:
new_df.to_csv('new_df.csv', index=False)
test = pd.read_csv('new_df.csv')

In [250]:
new_df.isnull().sum()

Extra_Staff        84913
author                 0
language               0
name                   0
narrator               0
price                  0
releasedate            0
time                   0
total_ratings       5822
version            86464
Average Ratings    72417
dtype: int64

In [252]:
test.isnull().sum()


Extra_Staff        84913
author                 0
language               0
name                   1
narrator               0
price                  0
releasedate            0
time                   0
total_ratings       5822
version            86464
Average Ratings    72417
dtype: int64

In [257]:
new_df[test['name'].isnull()]

Extra_Staff                 author language name  \
45761        None  Writtenby:SimoneKelly  English        

                             narrator price releasedate               time  \
45761  Narratedby:EbonyFord,K.Bernard   703    15-03-22  7 hrs and 34 mins   

       total_ratings  version  Average Ratings  
45761            0.0  [#1544]              NaN

In [258]:
df.iloc[45761]

name                                    #1544
author                  Writtenby:SimoneKelly
narrator       Narratedby:EbonyFord,K.Bernard
time                        7 hrs and 34 mins
releasedate                          15-03-22
language                              English
stars                           Not rated yet
price                                     703
Name: 45761, dtype: object

## Solving Dirty data issues

### Validity


In [259]:
# Dirt data: some values have strainge letters

# there is a possiblity that a value will be converted to nan if you encode it to ascii, so its better to check if any value is turning to nan

nan_created_values = new_df[new_df["name"].str.encode('ascii', 'ignore').isnull()]
new_df[new_df['name'].str.encode('ascii', 'ignore').isnull()] 

Empty DataFrame
Columns: [Extra_Staff, author, language, name, narrator, price, releasedate, time, total_ratings, version, Average Ratings]
Index: []

In [260]:
new_df[new_df["author"].str.encode('ascii', 'ignore').isnull()]

Empty DataFrame
Columns: [Extra_Staff, author, language, name, narrator, price, releasedate, time, total_ratings, version, Average Ratings]
Index: []

In [261]:
new_df[new_df["narrator"].str.encode('ascii', 'ignore').isnull()]

Empty DataFrame
Columns: [Extra_Staff, author, language, name, narrator, price, releasedate, time, total_ratings, version, Average Ratings]
Index: []

In [262]:
new_df['name'] = new_df["name"].str.encode('ascii', 'ignore').str.decode('ascii')
new_df["author"]   = new_df["author"].str.encode('ascii', 'ignore').str.decode('ascii')
new_df["narrator"] = new_df["narrator"].str.encode('ascii', 'ignore').str.decode('ascii')

In [263]:
# now remove the NaN values of the "name" column with the old ones
for index in nan_created_values.index:
    new_df["name"][index] = nan_created_values["name"][index]

In [264]:
new_df.isnull().sum()

Extra_Staff        84913
author                 0
language               0
name                   0
narrator               0
price                  0
releasedate            0
time                   0
total_ratings       5822
version            86464
Average Ratings    72417
dtype: int64

In [275]:
new_df['narrator'] = new_df['narrator'].replace('Narratedby:uncredited',np.nan)
new_df['narrator'] = new_df['narrator'].replace('Narratedby:anonymous',np.nan)

In [276]:
new_df.isnull().sum()

Extra_Staff        84913
author                 0
language               0
name                   0
narrator            1360
price                  0
releasedate            0
time                   0
total_ratings       5822
version            86464
Average Ratings    72417
dtype: int64

In [277]:
new_df['time']

0         2 hrs and 20 mins
1         13 hrs and 8 mins
2          2 hrs and 3 mins
3        11 hrs and 16 mins
4                    10 hrs
                ...        
87484     7 hrs and 34 mins
87485     10 hrs and 7 mins
87486     19 hrs and 4 mins
87487      4 hrs and 8 mins
87488       6 hrs and 1 min
Name: time, Length: 87489, dtype: object

In [317]:
def get_total_time(time):
    mins = 0
    hrs = 0
    if "mins" in time:
        mins = int(time.split("mins")[0].split(" ")[-2])
    if "hrs" in time:
        hrs = int(time.split("mins")[0].split("hrs")[-2].strip())

    return hrs*60 + mins


new_df['time'] = new_df['time'].apply(get_total_time)



In [318]:
new_df 

Extra_Staff                     author language  \
0            None  Writtenby:GeronimoStilton  English   
1            None      Writtenby:RickRiordan  English   
2            None       Writtenby:JeffKinney  English   
3            None      Writtenby:RickRiordan  English   
4            None      Writtenby:RickRiordan  English   
...           ...                        ...      ...   
87484        None     Writtenby:ChrisStewart  English   
87485        None    Writtenby:StephenO'Shea  English   
87486        None        Writtenby:MarkTwain  English   
87487        None   Writtenby:LaurenceSterne  English   
87488        None    Writtenby:MarkKurlansky  English   

                                      name                   narrator price  \
0                        Geronimo Stilton       Narratedby:BillLobely   468   
1                         The Burning Maze   Narratedby:RobbieDaymond   820   
2                             The Deep End      Narratedby:DanRussell   410   
3                     Daughter of the Deep  Narratedby:SoneelaNankani   615   
4      The Lightning Thief Percy Jackson,   Narratedby:JesseBernstein   820   
...                                    ...                        ...   ...   
87484            Last Days of the Bus Club    Narratedby:ChrisStewart   596   
87485                             The Alps      Narratedby:RobertFass   820   
87486                 The Innocents Abroad       Narratedby:FloGibson   938   
87487                A Sentimental Journey     Narratedby:AntonLesser   680   
87488                               Havana     Narratedby:FleetCooper   569   

               releasedate  time  total_ratings      version  Average Ratings  
0      2008-04-08 00:00:00   140           34.0  [#11 & #12]              5.0  
1      2018-01-05 00:00:00   788           41.0          NaN              4.5  
2      2020-06-11 00:00:00   123           38.0          NaN              4.5  
3      2021-05-10 00:00:00   676           12.0          NaN              4.5  
4      2010-01-13 00:00:00   600          181.0     [Book 1]              4.5  
...                    ...   ...            ...          ...              ...  
87484  2017-09-03 00:00:00   454            0.0          NaN              NaN  
87485             21-02-17   607            0.0          NaN              NaN  
87486             30-12-16  1144            0.0          NaN              NaN  
87487             23-02-11   248            0.0          NaN              NaN  
87488  2017-07-03 00:00:00   360            0.0          NaN              NaN  

[87489 rows x 11 columns]

In [331]:
new_df['releasedate'] = pd.to_datetime(new_df['releasedate'])

In [336]:
new_df['language'] = new_df['language'].str.capitalize()

In [338]:
new_df['language'].sample(10)

47287             English
23637             English
69450              French
69815              German
58805    Mandarin_chinese
25175             Swedish
48550             English
26568             English
82969             Italian
86001             English
Name: language, dtype: object

In [348]:
new_df['price'] = new_df['price'].replace({"Free":0})
new_df['price'] = new_df['price'].astype(np.int16)

In [349]:
new_df['price']

0        468
1        820
2        410
3        615
4        820
        ... 
87484    596
87485    820
87486    938
87487    680
87488    569
Name: price, Length: 87489, dtype: int16

## Solving Dirty data issues

### Accuracy


In [353]:
"Песочный человеÐabcº".encode('ascii', 'ignore').decode('ascii')

' abc'

In [355]:
new_df[new_df['name'] == ' ']

Extra_Staff      author  language name     narrator  price releasedate  \
394          None  Writtenby:   Russian       Narratedby:     99  2021-12-22   
396          None  Writtenby:   Russian       Narratedby:    200  2021-12-23   
433          None  Writtenby:   Russian       Narratedby:     99  2021-11-22   
434          None  Writtenby:   Russian       Narratedby:    132  2021-11-17   
440          None  Writtenby:   Russian       Narratedby:    132  2021-11-17   
...           ...         ...       ...  ...          ...    ...         ...   
87167        None  Writtenby:  Japanese       Narratedby:    837  2017-07-27   
87176        None  Writtenby:  Japanese       Narratedby:    139  2015-07-23   
87180        None  Writtenby:  Japanese       Narratedby:    139  2015-07-13   
87185        None  Writtenby:  Japanese       Narratedby:   1255  2015-12-20   
87193        None  Writtenby:  Japanese       Narratedby:    697  2018-05-16   

       time  total_ratings version  Average Ratings  
394      50            0.0     NaN              NaN  
396     244            0.0     NaN              NaN  
433     300            0.0     NaN              NaN  
434      46            0.0     NaN              NaN  
440      46            0.0     NaN              NaN  
...     ...            ...     ...              ...  
87167   453            0.0     NaN              NaN  
87176     0            0.0     NaN              NaN  
87180     0            0.0     NaN              NaN  
87185   423            0.0     NaN              NaN  
87193   348            0.0     NaN              NaN  

[373 rows x 11 columns]

In [357]:
new_df[new_df.duplicated(subset=['name', 'narrator','time','releasedate', 'language'])]


Extra_Staff       author  language  name      narrator  price  \
440          None   Writtenby:   Russian         Narratedby:    132   
1839         None   Writtenby:  Japanese         Narratedby:    139   
4749         None  Writtenby:,   Russian         Narratedby:     32   
4769         None  Writtenby:,   Russian         Narratedby:     32   
4772         None  Writtenby:,   Russian         Narratedby:     32   
...           ...          ...       ...   ...           ...    ...   
87182        None   Writtenby:  Japanese        Narratedby:,    279   
87186        None   Writtenby:  Japanese        Narratedby:,    279   
87189        None   Writtenby:  Japanese   /     Narratedby:    418   
87190        None   Writtenby:  Japanese  / ()   Narratedby:    418   
87192        None   Writtenby:  Japanese         Narratedby:    139   

      releasedate  time  total_ratings version  Average Ratings  
440    2021-11-17    46            0.0     NaN              NaN  
1839   2021-09-22    46            0.0     NaN              NaN  
4749   2021-12-23    11            0.0     NaN              NaN  
4769   2021-12-23    18            0.0     NaN              NaN  
4772   2021-12-23    11            0.0     NaN              NaN  
...           ...   ...            ...     ...              ...  
87182  2015-09-15    28            0.0     NaN              NaN  
87186  2016-04-01    31            0.0     NaN              NaN  
87189  2015-02-07    29            0.0     NaN              NaN  
87190  2016-12-04    31            0.0     NaN              NaN  
87192  2015-11-19     0            0.0     NaN              NaN  

[254 rows x 11 columns]

In [369]:
new_df.drop_duplicates(subset=['name', 'narrator','time','releasedate', 'language'],inplace = True)

In [366]:
new_df['author'] = new_df['author'].apply(lambda x: x.split('Writtenby:')[1]) 

In [367]:
new_df['author']

0        GeronimoStilton
1            RickRiordan
2             JeffKinney
3            RickRiordan
4            RickRiordan
              ...       
87484       ChrisStewart
87485      StephenO'Shea
87486          MarkTwain
87487     LaurenceSterne
87488      MarkKurlansky
Name: author, Length: 87489, dtype: object

In [12]:
#Nmae is not separated by white space instead they are merge together and first and last name are separated by capital letter
def foo(x):
    x = str(x)
    return re.sub(r'(?<=[a-z])(?=[A-Z])', ' ', x)
new_df['author'] = new_df['author'].apply(foo)
new_df['author']


0        Geronimo Stilton
1            Rick Riordan
2             Jeff Kinney
3            Rick Riordan
4            Rick Riordan
               ...       
87230       Chris Stewart
87231      Stephen O'Shea
87232          Mark Twain
87233     Laurence Sterne
87234      Mark Kurlansky
Name: author, Length: 87235, dtype: object